### PR TITLE
CW-266 verbose access controls for TOTP 2FA

### PR DESCRIPTION
**Pull request by Blazebrain**
Issue Number (if Applicable): Fixes #

# Description

Please include a summary of the changes and which issue is fixed / feature is added. 

# Pull Request - Checklist  

- [ ] Initial Manual Tests Passed
- [ ] Double check modified code and verify it with the feature/task requirements
- [ ] Format code
- [ ] Look for code duplication
- [ ] Clear naming for variables and methods

### CODE REVIEW

File: lib/src/screens/setup_2fa/modify_2fa_page.dart [hunk 1]
```diff
@@ -133,61 +130,61 @@ class _2FAControlsWidgetState extends State<_2FAControlsWidget>
                     .displayLarge!
                     .backgroundColor!,
                 tabs: [
-                  Tab(text: 'Narrow', height: 30),
-                  Tab(text: 'Normal', height: 30),
-                  Tab(text: 'Aggressive', height: 30),
+                  Tab(text: S.current.narrow, height: 30),
+                  Tab(text: S.current.narrow, height: 30),
```
Comment (SamsungGalaxyPlayer) [inline lib/src/screens/setup_2fa/modify_2fa_page.dart:134]: I assume this should be `s.current.normal`? The label in my app is `Narrow`

File: lib/src/screens/setup_2fa/modify_2fa_page.dart [hunk 2]
```diff
@@ -21,7 +22,44 @@ class Modify2FAPage extends BasePage {
...
+
+  @override
+  void dispose() {
+    super.dispose();
+    _tabController?.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Observer(builder: (context) {
```
Comment (OmarHatem28) [inline lib/src/screens/setup_2fa/modify_2fa_page.dart:61]: why rebuild the whole page when something is changed? just wrap each `SettingsSwitcherCell` in it's own observer

File: lib/src/screens/setup_2fa/modify_2fa_page.dart [hunk 3]
```diff
@@ -39,17 +77,124 @@ class Modify2FAPage extends BasePage {
...
+                    .primaryTextTheme
+                    .displayLarge!
+                    .backgroundColor!,
+                unselectedLabelColor: Theme.of(context)
+                    .primaryTextTheme
+                    .displayLarge!
+                    .backgroundColor!,
+                tabs: [
+                  Tab(text: S.current.narrow, height: 30),
+                  Tab(text: S.current.normal, height: 30),
```
Comment (OmarHatem28) [inline lib/src/screens/setup_2fa/modify_2fa_page.dart:134]: is there a specific reason why not directly using the already existing one? `SettingsChoicesCell`

**Review by OmarHatem28 — CHANGES_REQUESTED** [review]
Please ensure that you are using a line length of 100 and reformat the following pages

- [x] `settings_store.dart`
- [ ] `modify_2fa_page.dart`
- [ ] `di.dart`
- [ ] `exchange_page.dart`
- [ ] `set_up_2fa_viewmodel.dart`
- [ ] `send_view_model.dart`

and please fix the localization conflicts in favor of the main branch (i.e only add your this PR's related text but with the same format as the main branch)

File: lib/src/screens/wallet_list/wallet_list_page.dart [hunk 4]
```diff
@@ -163,15 +181,43 @@ class WalletListBodyState extends State<WalletListBody> {
...
+                    widget.authService.authenticateAction(
+                      context,
+                      route: Routes.newWalletType,
+                      conditionToDetermineIfToUse2FA: widget.walletListViewModel
+                          .shouldRequireTOTP2FAForCreatingNewWallets,
+                    );
+                  } else {
+                    Navigator.of(context).pushNamed(
+                      Routes.newWallet,
+                      arguments: widget.walletListViewModel.currentWalletType,
```
[the remark this answers is not in the document]
Comment (Blazebrain) [reply lib/src/screens/wallet_list/wallet_list_page.dart:218]: Yes it should be `Routes.newWalletType`, and without arguments

Switched it to it.

File: lib/store/settings_store.dart [hunk 5]
```diff
@@ -305,76 +433,132 @@ abstract class SettingsStoreBase with Store {
...
+        raw: sharedPreferences.getInt(PreferencesKey.selectedCake2FAPreset) ??
+            Cake2FAPresetsOptions.normal.raw);
+    final allowBiometricalAuthentication = sharedPreferences
+            .getBool(PreferencesKey.allowBiometricalAuthenticationKey) ??
+        false;
+    final shouldRequireTOTP2FAForAccessingWallet = sharedPreferences
+            .getBool(PreferencesKey.shouldRequireTOTP2FAForAccessingWallet) ??
+        false;
+    final shouldRequireTOTP2FAForSendsToContact = sharedPreferences
+            .getBool(PreferencesKey.shouldRequireTOTP2FAForSendsToContact) ??
```
[the remark this answers is not in the document]
Comment (Blazebrain) [reply lib/store/settings_store.dart]: Yes, missed that.

I've added them in now

File: res/values/strings_th.arb [hunk 6]
```diff
@@ -639,3 +651,9 @@
   "share": "แบ่งปัน",
   "slidable": "เลื่อนได้"
 }
+
+
+
+
+
+
```
[the remark this answers is not in the document]
Comment (Blazebrain) [reply res/values/strings_th.arb:661]: Alright

File: lib/view_model/send/send_view_model.dart [hunk 7]
```diff
@@ -193,9 +197,82 @@ abstract class SendViewModelBase with Store {
...
+    if (isContact) {
+      return shouldDisplayTOTP2FAForContact;
+    } else if (isInternalWallet) {
+      return shouldDisplayTOTP2FAForSendsToInternalWallet;
+    } else {
+      return shouldDisplayTOTP2FAForNonContact;
+    }
+  }
+
+  @action
```
[the remark this answers is not in the document]
Comment (Blazebrain) [reply lib/view_model/send/send_view_model.dart:254]: Alright, removed it

File: lib/src/screens/setup_2fa/modify_2fa_page.dart (hunk 2 above)
Comment (Blazebrain) [reply lib/src/screens/setup_2fa/modify_2fa_page.dart:61]: Yes, too costly 

Removed it and switched to individual Observers

File: lib/src/screens/setup_2fa/modify_2fa_page.dart (hunk 3 above)
Comment (Blazebrain) [reply lib/src/screens/setup_2fa/modify_2fa_page.dart]: Didn't know of the `SettingsChoicesCell` previously.

Switched this widget to it and adjusted the code.

File: lib/src/screens/contact/contact_list_page.dart [hunk 8]
```diff
@@ -26,95 +28,99 @@ class ContactListPage extends BasePage {
...
+            ),
+            ButtonTheme(
+              minWidth: 32.0,
+              height: 32.0,
+              child: TextButton(
+                  // FIX-ME: Style
+                  //shape: CircleBorder(),
+                  onPressed: () async {
+                    if (contactListViewModel
+                        .shouldRequireTOTP2FAForAddingContacts) {
```
[the remark this answers is not in the document]
Comment (Blazebrain) [reply lib/src/screens/contact/contact_list_page.dart:52]: Not really a duplicate check.

Normally, these routes are accessible without 2FA.

The 2FA implementation on it is to make sure that no one would be able to access these screens without the right access once 2fa has been activated.

If authentication fails when 2fa has been activated, it won't route to the screen.

When 2fa is not activated, users would still be able to carry out their normal activities.

File: lib/src/screens/dashboard/desktop_widgets/desktop_wallet_selection_dropdown.dart [hunk 9]
```diff
@@ -156,15 +156,30 @@ class _DesktopWalletSelectionDropDownState extends State<DesktopWalletSelectionD
...
+        arguments: widget.walletListViewModel.currentWalletType,
+        conditionToDetermineIfToUse2FA: widget
+            .walletListViewModel.shouldRequireTOTP2FAForCreatingNewWallets,
+      );
     } else {
-      Navigator.of(context).pushNamed(Routes.newWalletType);
+      widget._authService.authenticateAction(
+        context,
+        route: Routes.newWalletType,
+        arguments: widget.walletListViewModel.currentWalletType,
```
[the remark this answers is not in the document]
Comment (Blazebrain) [reply lib/src/screens/dashboard/desktop_widgets/desktop_wallet_selection_dropdown.dart:178]: Got it,

Removed it.

**Review by OmarHatem28 — APPROVED** [review]

**Pull request merged**